### DISCUSSION TITLE
Your agent's findings as a graph prior, and what emerges about uncertainty that the experiment log doesn't show.

### DISCUSSION THREAD

**Discussion by heidiEC** [karpathy/autoresearch#54] [Show and tell]
Following up on @karpathy's [post about asynchronous multi-agent collaboration](https://x.com/karpathy/status/...) and Discussion #43 from the Mar 5 autoresearch run.

The core problem: Discussion #43 is a flat list of 89 experiments: 15 kept, 74 discarded. It reads like a confident set of results. "Top 5 wins! Key themes! Dead ends!" But a second agent reading this has no way to know:

- How confident should I be in each finding?
- Which results need cross-validation most urgently?
- What did this agent *not* test that I should?
- Are these findings independent or do they interact?

We took the Mar 5 experiment log and encoded it as a **graph prior** — a lightweight structure (~1KB per finding) where each claim carries a certainty score, causal links to other claims, and a gap analysis that tells the next agent exactly what to prioritize.

## What the pod looks like

13 events extracted from 89 experiments. Here's the summary:

```
═══════════════════════════════════════════════════════════
  Karpathy Autoresearch Agent — Mar 5 Run
═══════════════════════════════════════════════════════════

  PATTERNS
    fixed-budget-bitter-lesson-inversion  value:90%  Z=0.30
    popular-defaults-dont-transfer        value:80%  Z=0.30

  CLAIMS (by certainty)
    throughput-over-params                Z=0.48  (strongest — 4+ confirming experiments)
    sliding-window-sweet-spot             Z=0.30
    rope-base-frequency-200k              Z=0.30
    load-bearing-components               Z=0.30
    label-smoothing-catastrophic          Z=0.30
    lr-schedule-stacking                  Z=0.30
    depth-width-razor-edge                Z=0.30
    swiglu-popularity-vs-performance      Z=0.30
    weight-decay-narrow-basin             Z=0.30
    seed-sensitivity-reproducibility-risk Z=0.30
    interaction-effects-untested          Z=0.30

  Events: 13 │ Avg Certainty: 0.31 │ Need Validation: 13
═══════════════════════════════════════════════════════════
```

## The causal structure

The flat log doesn't show relationships between findings. The graph prior does:

```mermaid
graph TD
    classDef pattern fill:#e8f5e9,stroke:#1b5e20
    classDef claim fill:#f3e5f5,stroke:#6a1b9a

    claim_throughput[throughput-over-paramsZ:0.48]:::claim
    claim_swiglu[swiglu-popularity-vs-performanceZ:0.30]:::claim
    claim_label[label-smoothing-catastrophicZ:0.30]:::claim
    claim_depth[depth-width-razor-edgeZ:0.30]:::claim
    claim_seed[seed-sensitivity-reproducibilityZ:0.30]:::claim
    claim_interaction[interaction-effects-untestedZ:0.30]:::claim
    pattern_bitter[fixed-budget-bitter-lesson-inversionZ:0.30]:::pattern
    pattern_transfer[popular-defaults-dont-transferZ:0.30]:::pattern

    claim_throughput ==>|ENABLES| pattern_bitter
    claim_swiglu -->|BUILDS_ON| pattern_transfer
    claim_label -->|BUILDS_ON| pattern_transfer
    claim_depth -->|BUILDS_ON| claim_interaction
    claim_seed -->|BUILDS_ON| claim_interaction
```

Two emergent patterns surface that the experiment log mentions informally but doesn't track:

1. **Fixed-budget Bitter Lesson inversion**: At fixed compute, simplicity beats complexity. The throughput-over-params finding (the single biggest win) *enables* this broader insight. Every param added steals a training step.

2. **Popular defaults don't transfer**: SwiGLU (no improvement), cosine warmdown (worse than linear), label smoothing (catastrophic), MQA (much worse). Techniques validated at scale fail under throughput constraints. The compute budget creates a different optimization landscape.

## What the gap analysis reveals

This is the part the flat log can't do. Every finding ranked by `value × (1 - certainty)` — high-signal, low-confidence items first:

```
🔴 HIGH PRIORITY — All 13 findings need validation

  label-smoothing-catastrophic     gap=0.997  ← Highest priority
    single source, 1 observation, very low certainty
    
  sliding-window-sweet-spot        gap=0.945
    single source, 1 observation, very low certainty
    
  load-bearing-components          gap=0.945
    single source, 1 observation, very low certainty
    
  throughput-over-params           gap=0.780
    ← Strongest certainty (Z=0.48) because 4+ confirming
    experiments within this run. Still single-source.
    
  interaction-effects-untested     gap=0.630
    ← The meta-gap: greedy sequential search assumed
    independence. Seed sensitivity suggests it shouldn't.
```

**The brutal truth: all 13 findings are single-source, single-run, single-seed, single-GPU.** The experiment log doesn't flag this. The pod does.

## What the next agent should do

The gap analysis IS the coordination mechanism. A second agent reading this pod knows:

1. **Cross-validate label smoothing** (gap=0.997) — Is +0.34 BPB reproducible, or was something else wrong? This is the strongest claim by value but the least validated.

2. **Test sliding window on different hardware/data** (gap=0.945) — Does SSSSL with 1/8 window hold on A100? On different datasets? At different budgets?

3. **Run interaction experiments** (gap=0.630) — The agent searched one variable at a time. The spiky optima (depth 9 not 10, window 1/8 not 1/16, RoPE 200K not 300K) suggest hyperparameters interact. Nobody tested this.

4. **Multi-seed validation** (gap=0.525) — Seed 42→137 flipped a "kept" result. How many of the 15 kept experiments survive across seeds?

A second agent doesn't re-run 89 experiments. It reads the gap analysis and runs the 5-10 experiments that would move the highest-priority claims from Z=0.30 to Z=0.5+. A third agent does the same. Certainty accumulates across the community without centralized coordination.

## The collaboration primitive

This is what we think multi-agent research coordination looks like:

- **Each agent gets a pod** — a graph prior with certainty-scored findings
- **Pods don't merge** — agents hold divergent beliefs. Convergence happens when evidence agrees. Contradictions are the most valuable signal.
- **Gap analysis replaces coordination** — no central planner needed. The certainty landscape tells each agent where to focus.
- **~1KB per finding propagates** — not code, not experiment logs, not 89-row tables. The *learning* is what travels.

The raw experiment log from Discussion #43 is 89 rows of deltas. The pod is 13 causal claims with certainty, links, and gap priorities. A second agent can read the pod in milliseconds and know exactly what to validate next.

Git tracks what was done. A graph prior tracks what was *learned* — and what still needs to be learned.

---

**Pod source**: [github.com/Nervous-Machine/context-pod](https://github.com/Nervous-Machine/context-pod) — see [MULTI_AGENT.md](https://github.com/Nervous-Machine/context-pod/blob/main/MULTI_AGENT.md) for the full architecture.

**Live pod**: The autoresearch-mar5 pod is running on our MCP server. The same protocol works for people, agents, and physical systems (we have a separate instance mapping orbital physics with satellite data).

We're not proposing this as the only approach — but the experiment log from Discussion #43 was a perfect test case for what graph priors add to multi-agent research. Happy to answer questions or help anyone encode their agent's runs.

— Heidi Bennett, [Nervous Machine](https://nervousmachine.com)

**Comment by guarinoindustries-ctrl** [comment]
Andrej didn't come up with this idea originally, this guy did: https://github.com/topherchris420/james_library
